omnigibson/object_states/contact_particles.py

```diff
@@ -67,23 +74,38 @@ def cache_info(self, get_value_args):
         # Run super first
         info = super().cache_info(get_value_args=get_value_args)
 
-        # Store the system's number of particles for each instancer
+        # Store the system's particle positions for each instancer
         for arg in get_value_args:
             if inspect.isclass(arg) and issubclass(arg, PhysicalParticleSystem):
-                info[arg] = arg.n_particles
+                info[arg] = {instancer: instancer.particle_positions for instancer in arg.particle_instancers.values()}
 
         return info
 
     def _cache_is_valid(self, get_value_args):
         # Run super first
         is_valid = super()._cache_is_valid(get_value_args=get_value_args)
 
-        # If it's valid, do final check with system
-        if is_valid:
-            for arg, info in self._cache[get_value_args]["info"].items():
-                if inspect.isclass(arg) and issubclass(arg, PhysicalParticleSystem):
-                    # Make sure the number of particles in the system is the same; otherwise
-                    # something has changed, so we need to update the cache internally
-                    is_valid = arg.n_particles == info
+        if not is_valid:
+            return False
+
+        for arg, info in self._cache[get_value_args]["info"].items():
+            if inspect.isclass(arg) and issubclass(arg, PhysicalParticleSystem):
+                # TODO: adopt the has_changed mechanism in object_state_base
+                # Check if the particle positions have changed
+
+                # If the instancers don't match, return False
+                if list(arg.particle_instancers.values()) != list(info.keys()):
+                    return False
+
+                # If there are no instancers, skip
+                if len(info.keys()) == 0:
+                    continue
+
+                arg_pos= np.vstack([instancer.particle_positions for instancer in arg.particle_instancers.values()])
+                info_pos = np.vstack([particle_positions for particle_positions in info.values()])
+
+                # If any of the particles moved, return False
+                if np.any(np.linalg.norm(arg_pos - info_pos, axis=1) >= m.POSITIONAL_VALIDATION_EPSILON):
```

Comment:
I'm guessing this check is indeed faster than just not caching anything?

Reply:
I am actually not sure. This check is slightly slower than 
```
inbound_idxs = ((lower < positions) & (positions < upper)).all(axis=-1).nonzero()[0]
```

So if no particles in inbound to further test with `get_physx_scene_query_interface`, then this check will actually make things slower.

Reply:
Commented out for now.